pkg/kubelet/deviceplugin/manager.go

```diff
@@ -212,21 +213,38 @@ func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
 		return
 	}
 
-	stream, err := e.list()
-	if err != nil {
-		glog.Errorf("Failed to List devices for plugin %v: %v", r.ResourceName, err)
-		return
-	}
+	closeChan := make(chan bool)
+	go e.run(closeChan)
 
 	go func() {
-		e.listAndWatch(stream)
-
-		m.mutex.Lock()
-		if old, ok := m.Endpoints[r.ResourceName]; ok && old == e {
-			delete(m.Endpoints, r.ResourceName)
+		// TODO: Consider making timeout value configurable.
+		t := time.NewTimer(5 * time.Second)
+		defer t.Stop()
+
+		for {
+			select {
+			case <-t.C:
```

Comment:
I think you should just wait for the time to tick and check at the end if there are devices or not that would remove the need for a channel

Reply:
How about:
```
<-time.Affter(time.Second * 2)
```

Reply:
But this way, it would block until timeout, and we could not handle in time if listWatch of endpoint exits during this period

Reply:
I was still commenting in the context of the goroutine :)
Thought about this a bit more here is what I think might be a cleaner solution.

```golang
go func() {
	<-time.After(5 * time.Second)

	e.mutex.Lock()
	nDevices = len(e.devices)
	e.mutex.Unlock()

	if nDevices == 0 {
		e.stop()
	}

	select {
		case <-e.stopChan:
			e.stop()
		case <-e.exitChan:
			removeEndpoint(e)
			return
	}
}

func (e *endpoint) stop() {
	e.cancel()
}
```

And you would change Endpoint.runlike this:
```
func (e *endpoint) run() {
     defer close(e.exitChan)
}```

Reply:
The snippet is missing a for loop before the select